File: wc-clone/wc-javascript/wc.js
import fs from "node:fs";
import process from "node:process";
import yargs from 'yargs';

function countChars(filePath) {
    try {
        const data = fs.readFileSync(filePath, 'utf8');
        return Array.from(data).length;
    } catch (error) {
        if (error.code === 'ENOENT') {
            return null;
        } 
        throw error;
    }
}

function countBytes(filePath) {
    try {
        const data = fs.readFileSync(filePath);
        return data.length;
    } catch (error) {
        if (error.code === 'ENOENT') {
            return null;
        } 
        throw error;
    } 
} 

function countLines(filePath) {
    try {
        const data = fs.readFileSync(filePath, 'utf8');
        const lines = data.split('\n');
        return lines[lines.length - 1] === '' ? lines.length - 1 : lines.length;
    } catch(error) {
        if (error.code === 'ENOENT') {
            return null;
        } 
        throw error;
    } 
}

function countWords(filePath) {
    try {
        const data = fs.readFileSync(filePath, 'utf8');
        const words = data.trim().split(/\s+/);
        return words.length;
    } catch (error) {
        if (error.code === 'ENOENT') {
            return null;
        }
        throw error;
    }
}

function countLinesFromInput(inputData) {
    const lines = inputData.split('\n');
    return lines[lines.length - 1] === '' ? lines.length - 1 : lines.length;
}

function countWordsFromInput(inputData) {
    const words = inputData.trim().split(/\s+/);
    return words.length;
}

function countBytesFromInput(inputData) {
    return Buffer.byteLength(inputData, 'utf8');
} 

async function runWordCount(argv = process.argv) {
    const args = yargs(argv.slice(2))
        .scriptName('wc')
        .usage('Usage: $0 [options] <file>')
        .option('c', {
            alias: 'bytes',
            describe: 'Count bytes',
            type: 'boolean',
        })
        .option('l', {
            alias: 'lines',
            describe: 'Count lines',
            type: 'boolean',
        })
        .option('w', {
            alias: 'words',
            describe: 'Count words',
            type: 'boolean',
        })
        .option('m', {
            alias: 'chars',
            describe: 'Count characters',
            type: 'boolean',
        })
        .help('h')
        .alias('h', 'help')
        .argv;

    const filePath = args._[0];
    let inputData = '';

    if (!filePath) {
        process.stdin.setEncoding('utf8');
        for await (const chunk of process.stdin) {
            inputData += chunk;
        }
    }

    if(!args.chars && !args.bytes && !args.lines && !args.words) {
        const lineCount = filePath ? countLines(filePath) : countLinesFromInput(inputData);
        const wordCount = filePath ? countWords(filePath) : countWordsFromInput(inputData);
        const byteCount = filePath ? countBytes(filePath) : countBytesFromInput(inputData);

        if (lineCount === null || wordCount === null || byteCount === null) {
            console.error(`Error: Could not '${filePath ? `open file '${filePath}'` : 'read from standard input'}`); 
            process.exit(1);
        } 

        const result = `${lineCount} ${wordCount} ${byteCount} ${filePath ? `${filePath}` : ''}`.trim();
        console.log(result);
        return result;
    } 

    if (argv.chars) {
        const count = countChars(filePath);
        if (count === null) {
            console.error(`Error: Could open file '${filePath}'`);
            process.exit(1);
        } 
        console.log(`${count} ${filePath}`);
    } else if (argv.bytes) {
        const count = countBytes(filePath);
        if (count === null) {
            console.error(`Error: Could not open file '${filePath}'`);
            process.exit(1);
        } 
        console.log(`${count} ${filePath}`);
    } else if (argv.lines) {
        const count = countLines(filePath);
        if (count === null) {
            console.error(`Error: Could not open file '${filePath}'`);
            process.exit(1);
        }
        console.log(`${count} ${filePath}`);
    } else if (argv.words) {
        const count = countWords(filePath);
        if (count === null) {
            console.error(`Error: Could not open file '${filePath}'`);
            process.exit(1);
        }
        console.log(`${count} ${filePath}`);
    } else {
        console.error('Error: Missing -c or -l flag');
        process.exit(1);
    } 
}

if (import.meta.url === `file://${process.argv[1]}`) {
    runWordCount();
}


export {
    countBytes,
    countLines,
    countWords,
    countChars,
    countBytesFromInput,
    countWordsFromInput,
    countLinesFromInput,
    runWordCount,
}; 

File: wc-clone/wc-python/wc.py
import argparse 
import io
import sys


def count_bytes(file):
    if file is sys.stdin:
        content = file.read().encode('utf-8')
        return len(content)
    else:
        try:
            with open(file, "rb") as f:
                content = f.read()
                return len(content)
        except FileNotFoundError:
            return None


def count_chars(file):
    if file is sys.stdin:
        content = file.read()
        return len(content)
    else:
        try:
            with open(file, "r", encoding="utf-8") as f:
                content = f.read()
                return len(content)
        except FileNotFoundError:
            return None

def count_lines(file):
    if file is sys.stdin:
        return sum(1 for _ in file)
    else:
        try:
            with open(file, "r") as f:
                return sum(1 for _ in f)
        except FileNotFoundError:
            return None

def count_words(file):
    if file is sys.stdin:
        content = file.read()
        words = content.split()
        return len(words)
    else:
        try:
            with open(file, "r") as f:
                content = f.read()
                words = content.split()
                return len(words)
        except FileNotFoundError:
            return None

def default_count(file):
    try:
        if file is sys.stdin:
            content = file.read()
            line_count = content.count('\n')
            if content and not content.endswith('\n'):
                line_count += 1
            word_count = len(content.split())
            char_count = len(content)
        else:
            with open(file, "r") as f:
                content = f.read()
                line_count = content.count('\n')
                if content and not content.endswith('\n'):
                    line_count += 1
                word_count = len(content.split())
                char_count = len(content)
        return line_count, word_count, char_count
    except FileNotFoundError:
        return None, None, None


def main():
    parser = argparse.ArgumentParser(description="Word Count")
    parser.add_argument("-c", "--bytes", action="store_true", help="Count bytes")
    parser.add_argument("-l", "--lines", action="store_true", help="Count lines")
    parser.add_argument("-w", "--words", action="store_true", help="Count words")
    parser.add_argument("-m", "--chars", action="store_true", help="Count characters")
    parser.add_argument("file", nargs="?", default=None, help="Input file")
    args = parser.parse_args()

    if args.file == "-" or args.file is None:
        file = sys.stdin
    else:
        file = args.file

    if not any([args.bytes, args.lines, args.words, args.chars]):
        line_count, word_count, char_count = default_count(file)
        if line_count is None:
            print(f"Error: could not open file '{args.file}'")
            sys.exit(1)
        print(f"{line_count} {word_count} {char_count} {args.file}")
    else:
        if args.chars:
            count = count_chars(file)
            print(f"{count}")
        elif args.bytes:
            count = count_bytes(file)
            print(f"{count}")
        elif args.lines:
            count = count_lines(file)
            print(f"{count}")
        elif args.words:
            count = count_words(file)
            print(f"{count}")
        else:
            print("Error: Missing -c or -l, or -w flag")
            sys.exit(1)


if __name__ == "__main__":
    main()

